[mlir][debug] Handle DIImportedEntity.

abidh opened the pull request:
The `DIImporedEntity` can be used to represent imported entities like C++'s namespace with using directive or fortran's moudule with use statement.

This PR adds `DIImportedEntityAttr` and 2-way translation from `DIImportedEntity` to `DIImportedEntityAttr` and vice versa.

When an entity is imported in a function, the `retainedNodes` field of the `DISubprogram` contains all the imported nodes. See the C++ code and the LLVM IR below.

```
void test() {
    using namespace n1;
 ...
}

!2 = !DINamespace(name: "n1", scope: null)
!16 = distinct !DISubprogram(name: "test", ..., retainedNodes: !19) !19 = !{!20}
!20 = !DIImportedEntity(tag: DW_TAG_imported_module, scope: !16, entity: !2 ...)
```

This PR makes sure that the translation from mlir to `retainedNodes` field happens correctly both ways.

There was no obvious place where I could place the `DIImportedEntityAttr` in the mlir. I have decided to piggy back on the `FusedLoc` of `FuncOp` which is already used to carry `DISubprogramAttr`. I am open to other ideas here.

This PR currently does not handle entities imported in a global scope but that should be easy to handle in a subsequent PR.

abidh committed:
[mlir][debug] Handle DIImportedEntity.

The `DIImporedEntity` can be used to represent imported entities like
C++'s namespace with using directive or fortran's moudule with use
statement.

This PR adds `DIImportedEntityAttr` and 2-way translation from
`DIImportedEntity` to `DIImportedEntityAttr` and vice versa.

When an entity is imported in a function, the `retainedNodes` field of
the `DISubprogram` contains all the imported nodes. See the C++ code and the
LLVM IR below.

void test() {
    using namespace n1;
 ...
}

!2 = !DINamespace(name: "n1", scope: null)
!16 = distinct !DISubprogram(name: "test", ..., retainedNodes: !19)
!19 = !{!20}
!20 = !DIImportedEntity(tag: DW_TAG_imported_module, scope: !16, entity: !2 ...)

This PR makes sure that the translation from mlir to `retainedNodes`
field happens correctly both ways.

There was no obvious place where I could place the `DIImportedEntityAttr`
in the mlir. I have decided to piggy back on the `FusedLoc` of `FuncOp`
which is already used to carry `DISubprogramAttr`. I am open to other
ideas here.

This PR currently does not handle entities imported in a global scope
but that should be easy to handle in a subsequent PR.

Dinistro commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -619,6 +619,25 @@ def LLVM_DINamespaceAttr : LLVM_Attr<"DINamespace", "di_namespace",
+                                           /*traits=*/[], "DINodeAttr"> {
+  let parameters = (ins
+    LLVM_DITagParameter:$tag,
+    OptionalParameter<"DIScopeAttr">:$scope,
+    OptionalParameter<"DINodeAttr">:$entity,
+    OptionalParameter<"DIFileAttr">:$file,
+    OptionalParameter<"unsigned">:$line,
+    OptionalParameter<"StringAttr">:$name,
+    OptionalArrayRefParameter<"DINodeAttr">:$elements
```
Does LLVM really make all these fields optional?

Dinistro commented on mlir/test/Target/LLVMIR/Import/debug-info.ll:
```diff
@@ -792,3 +792,29 @@ define void @string_type(ptr %arg1) {
+
+; CHECK-DAG: #[[M:.+]] = #llvm.di_module<{{.*}}name = "mod1"{{.*}}>
+; CHECK-DAG: #[[SP:.+]] = #llvm.di_subprogram<{{.*}}name = "imp_fn"{{.*}}>
+; CHECK-DAG: llvm.di_imported_entity{{.*}}tag = DW_TAG_imported_module, scope = #[[SP]], entity = #[[M]]
```
Can you check that this attribute is attached to the expected place in the IR?

Dinistro reviewed (commented):
Is there any restriction on the `scope` of a `di_imported_entity`?
In general, I wonder if it would be better to attach them to the DISubprogramAttr instead, but that leads to yet another way one can get a cyclic attribute (cc @zyx-billy).

abidh commented:
> Is there any restriction on the `scope` of a `di_imported_entity`? In general, I wonder if it would be better to attach them to the DISubprogramAttr instead, but that leads to yet another way one can get a cyclic attribute (cc @zyx-billy).

I am not aware of any restriction on the scope. I think it can take any `DIScope` but in most cases it would be `DISubprogram`, `DIFile` or `DICompileUnit`. 

Please correct me if I am wrong but I thought Attributes are sort of readonly once created. So adding a field in `DISubprogramAttr` to hold them would not work as

1. `DIImportedEntity` needs scope so it can only be created once a parent like `DISubprogramAttr` is available.
2. But once `DISubprogramAttr` has been created, we can't later modify it to set a list of imported entities in it.

Similarly, we can't attach attribute or location to `DISubprogramAttr` so attaching list of `DIImportedEntity` to `FuncOp` seemed like the only option to me.

gysit commented:
I agree that it would be desired to replicate LLVM's structure and add the new attribute using a retained nodes field in the subprogram.

> Please correct me if I am wrong but I thought Attributes are sort of readonly once created. So adding a field in DISubprogramAttr to hold them would not work as

There is a mechanism to do this in principle but it is only implement for DICompositeType so far. Essentially DISubprogram would have to implement the LLVM_DIRecursiveTypeAttrInterface (https://github.com/llvm/llvm-project/blob/0160d817c2c1bd3f32986d9cf8b938e613ac3d12/mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td#L374). When translating you would then start translating the subprogram retained nodes but you remember that the traversal already saw the subprogram once and give it an id. So the DIImportedEntity scope would only contain an id instead of the full subprogram. Once all nested attributes are translated you then create the actual subprogram which has the id attached.

I believe we will need some consulting from @zyx-billy to get this working though (the mechanism is non-trivial). In principle the import of debug info already has the mechanism working but only for DICompositeType.

gysit commented:
https://discourse.llvm.org/t/handling-cyclic-dependencies-in-debug-info/67526/17 discusses the topic. I think it may make sense to look at the last posts of Billy and look at the relevant PRs.

https://github.com/llvm/llvm-project/pull/80251
https://github.com/llvm/llvm-project/pull/85850
https://github.com/llvm/llvm-project/pull/87295

zyx-billy commented:
Thank you @gysit for the overview. Yes, unfortunately this is our only solution for cyclic DI metadata right now. Coincidentally we were just discussing in #98203 how we may need to add recursion support so that DISubprogram may be used to "prune" cycles, and once that's added this should work automatically, but neither of us got to it yet...

Another solution, if an ImportedEntity is very limited in where it can exist, is to forgo the back-referencing "scope" field from the ImportedEntity and just rely on the context when exporting back to llvm to assign it the scope again. Is there such a guarantee on where an ImportedEntity can appear?

abidh commented:
Thanks @gysit @zyx-billy and @Dinistro for your comments.

@zyx-billy the idea to remove the scope field from the imported entity is interesting one. This seems like something that could be done as scope could be inferred depending on where these entities are stores. I will go and experiment on it.

abidh committed:
[mlir][debug] Add retainedNodes field in DISubprogramAttr.

As discussed in the comments, this commit adds retainedNodes field in
the DISubprogramAttr. To remove the cyclic dependency, the scope
field has been removed from DIImportedEntity as it can be inferred
depending on the contained of the entities.

abidh commented:
Hi @zyx-billy, @gysit. @Dinistro   
I pushed a commit that removes scope from DIImportedEntity. I also added a retainedNodes field in the DISubprogramAttr to better mimic DISubprogram. There were some small changes to handle comments from @Dinistro. Please have a look and let me know if this approach makes sense.

Dinistro commented on mlir/lib/Dialect/LLVMIR/Transforms/DIScopeForLLVMFuncOp.cpp:
```diff
@@ -79,7 +79,7 @@ static void addScopeToFunction(LLVM::LLVMFuncOp llvmFunc,
       fileAttr,
       /*line=*/line,
-      /*scopeline=*/col, subprogramFlags, subroutineTypeAttr);
+      /*scopeline=*/col, subprogramFlags, subroutineTypeAttr, {});
```
Nit: Also add the param name, like it's done for `line` and `scopeline` here.

abidh replied:
Done.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -208,6 +208,21 @@ DINamespaceAttr DebugImporter::translateImpl(llvm::DINamespace *node) {
                               node->getExportSymbols());
 }
 
+DIImportedEntityAttr
```
This requires a clear TODO comment that we drop this information that also explains why it is dropped. Ultimately, we would want to import even the cyclic structures, but this can wait for now.

gysit replied:
Maybe add a brief comment on the attribute itself why there is no scope field.

abidh replied:
I added a comment on the attribute.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -306,6 +306,18 @@ llvm::DISubprogram *DebugTranslation::translateImpl(DISubprogramAttr attr) {
 
+  SmallVector<llvm::Metadata *> retainedNodes;
+
+  for (auto nodeAttr : attr.getRetainedNodes()) {
```
Nit: Do not use auto here.

abidh replied:
Thanks for catching this. Done.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -306,6 +306,18 @@ llvm::DISubprogram *DebugTranslation::translateImpl(DISubprogramAttr attr) {
       static_cast<llvm::DISubprogram::DISPFlags>(attr.getSubprogramFlags()),
       compileUnit);
 
+  SmallVector<llvm::Metadata *> retainedNodes;
```
Nit: Add a comment on why this is done in a delayed manner.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugTranslation.h:
```diff
@@ -89,6 +89,7 @@ class DebugTranslation {
   llvm::DISubrange *translateImpl(DISubrangeAttr attr);
   llvm::DISubroutineType *translateImpl(DISubroutineTypeAttr attr);
   llvm::DIType *translateImpl(DITypeAttr attr);
+  llvm::DIImportedEntity *translate(DIImportedEntityAttr attr, llvm::DIScope *);
```
Nit: This needs a comment, given that it does something non-obvious.

abidh replied:
Done.

Dinistro reviewed (commented):
Dropped some nit comments, mainly regarding missing documentation that will be required for a case that diverges from LLVM's modeling.

gysit commented on mlir/include/mlir-c/Dialect/LLVM.h:
```diff
@@ -316,7 +316,8 @@ MLIR_CAPI_EXPORTED MlirAttribute mlirLLVMDISubprogramAttrGet(
     MlirAttribute scope, MlirAttribute name, MlirAttribute linkageName,
     MlirAttribute file, unsigned int line, unsigned int scopeLine,
-    uint64_t subprogramFlags, MlirAttribute type);
+    uint64_t subprogramFlags, MlirAttribute type, intptr_t nNodes,
+    MlirAttribute const *nodes);
```
```suggestion
    uint64_t subprogramFlags, MlirAttribute type, intptr_t nRetainedNodes,
    MlirAttribute const *retainedNodes);
```
nit: The name should probably match the attribute fields?

abidh replied:
Done.

gysit commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -306,6 +306,18 @@ llvm::DISubprogram *DebugTranslation::translateImpl(DISubprogramAttr attr) {
+  SmallVector<llvm::Metadata *> retainedNodes;
+
+  for (auto nodeAttr : attr.getRetainedNodes()) {
+    if (DIImportedEntityAttr importedAttr =
```
```suggestion
    if (auto importedAttr =
```

abidh replied:
Done.

gysit commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -223,11 +238,18 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
   if (node->getType() && !type)
     return nullptr;
+
+  SmallVector<DINodeAttr> retainedNodes;
+
+  for (auto node : node->getRetainedNodes())
```
```suggestion
  SmallVector<DINodeAttr> retainedNodes;
  for (auto node : node->getRetainedNodes())
```

abidh replied:
Done.

gysit commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -208,6 +208,21 @@ DINamespaceAttr DebugImporter::translateImpl(llvm::DINamespace *node) {
 
+DIImportedEntityAttr
+DebugImporter::translateImpl(llvm::DIImportedEntity *node) {
+  SmallVector<DINodeAttr> elements;
+
+  for (llvm::DINode *element : node->getElements()) {
```
```suggestion
  SmallVector<DINodeAttr> elements;
  for (llvm::DINode *element : node->getElements()) {
```

abidh replied:
Done.

gysit reviewed (commented):
Added some more nit comments. The approach seems ok for now!

abidh committed:
Handle review comments.

Mostly formatting nits and adding comments.

abidh committed:
Fix flang build.

Add a default value for retainedNodes field.

abidh commented:
Thanks @Dinistro and @gysit for your comments. I think I have handled all the nits.

The flang build was failing due to extra field in DISubprogramAttr. I have updated it to make the build pass. Please let me know if that needs to go as separate PR.

gysit commented:
Fixing the build in this PR is fine.

pawosm-arm commented:
Is this related to issue #98883?

abidh commented:
> Is this related to issue #98883?

Yes, this is mlir side of the work. Once this is in then I will open a PR for flang side of the fix.

abidh committed:
Fix some typos.

The function parameter names were swapped in 2 places.

abidh commented:
Hi @gysit, @Dinistro 

Any further comments on this PR?

Dinistro commented:
Cannot check right now, but if @gysit is happy, I'm happy.
anote: I suggest to re-request review in such cases, then it shown up in the list of open reviews

gysit reviewed (approved):
LGTM

definelicht commented:
@abidh We're are experiencing a crash in the cyclic import cache (cc @zyx-billy) after this change.
```
mlir-translate: llvm-project/mlir/include/mlir/Support/CyclicReplacerCache.h:191: typename CyclicReplacerCache<InT, OutT>::CacheEntry mlir::CyclicReplacerCache<llvm::DINode *, mlir::LLVM::DINodeAttr>::lookupOrInit(InT) [InT = llvm::DINode *, OutT = mlir::LLVM::DINodeAttr]: Assertion `it->second.size() <= 2 && "illegal 3rd repeat of input"' failed.
```
I'm unsure if this warrants a revert, since this might be a bug uncovered in the caching code? (cc @Dinistro and @gysit). Attaching a reproducer below:

```
; ModuleID = 'reproducer.ll'
source_filename = "reproducer.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define void @foo() !dbg !23 {
  ret void
}

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!22}

!0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !1, producer: "clang", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug, enums: !2, retainedTypes: !2, globals: !3, imports: !2, splitDebugInlining: false, nameTableKind: None)
!1 = !DIFile(filename: "/my/dir/foo.cpp", directory: "/my/dir")
!2 = !{}
!3 = !{!4, !11, !17}
!4 = !DIGlobalVariableExpression(var: !5, expr: !DIExpression())
!5 = distinct !DIGlobalVariable(name: "foo", scope: !0, file: !6, line: 110, type: !7, isLocal: false, isDefinition: true)
!6 = !DIFile(filename: "dir/foo.cpp", directory: "dir")
!7 = distinct !DICompositeType(tag: DW_TAG_class_type, name: "unordered_map<unsigned int, std::__1::thread, std::__1::hash<unsigned int>, std::__1::equal_to<unsigned int>, std::__1::allocator<std::__1::pair<const unsigned int, std::__1::thread> > >", scope: !9, file: !8, line: 1103, size: 320, flags: DIFlagTypePassByReference | DIFlagNonTrivial, elements: !2, templateParams: !2, identifier: "_ZTSNSt3__113unordered_mapIjNS_6threadENS_4hashIjEENS_8equal_toIjEENS_9allocatorINS_4pairIKjS1_EEEEEE")
!8 = !DIFile(filename: "include/c++/v1/unordered_map", directory: "/usr/")
!9 = !DINamespace(name: "__1", scope: !10, exportSymbols: true)
!10 = !DINamespace(name: "std", scope: null)
!11 = !DIGlobalVariableExpression(var: !12, expr: !DIExpression())
!12 = distinct !DIGlobalVariable(scope: null, file: !13, line: 252, type: !14, isLocal: true, isDefinition: true)
!13 = !DIFile(filename: "include/c++/v1/__thread/thread.h", directory: "/usr/")
!14 = !DICompositeType(tag: DW_TAG_array_type, baseType: !15, size: 208, elements: !2)
!15 = !DIDerivedType(tag: DW_TAG_const_type, baseType: !16)
!16 = !DIBasicType(name: "char", size: 8, encoding: DW_ATE_signed_char)
!17 = !DIGlobalVariableExpression(var: !18, expr: !DIExpression())
!18 = distinct !DIGlobalVariable(name: "piecewise_construct", linkageName: "_ZNSt3__119piecewise_constructE", scope: !9, file: !19, line: 23, type: !20, isLocal: false, isDefinition: true)
!19 = !DIFile(filename: "include/c++/v1/__utility/piecewise_construct.h", directory: "/usr/")
!20 = !DIDerivedType(tag: DW_TAG_const_type, baseType: !21)
!21 = distinct !DICompositeType(tag: DW_TAG_structure_type, name: "piecewise_construct_t", scope: !9, file: !19, line: 20, size: 8, flags: DIFlagTypePassByValue, elements: !2, identifier: "_ZTSNSt3__121piecewise_construct_tE")
!22 = !{i32 2, !"Debug Info Version", i32 3}
!23 = distinct !DISubprogram(name: "foo", scope: !6, file: !6, line: 97, type: !24, scopeLine: 98, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !26)
!24 = distinct !DISubroutineType(types: !25)
!25 = !{null}
!26 = !{!27}
!27 = !DILocalVariable(name: "foo", arg: 1, scope: !23, file: !6, line: 97, type: !28)
!28 = !DIDerivedType(tag: DW_TAG_typedef, name: "foo_t", file: !29, line: 14, baseType: !30)
!29 = !DIFile(filename: "dir/foo.h", directory: "dir")
!30 = !DIDerivedType(tag: DW_TAG_typedef, name: "uint32_t", file: !31, line: 26, baseType: !32)
!31 = !DIFile(filename: "include/bits/stdint-uintn.h", directory: "/usr/")
!32 = !DIDerivedType(tag: DW_TAG_typedef, name: "__uint32_t", file: !33, line: 42, baseType: !34)
!33 = !DIFile(filename: "include/bits/types.h", directory: "/usr/")
!34 = !DIBasicType(name: "unsigned int", size: 32, encoding: DW_ATE_unsigned)
```
